core/src/test/java/com/linecorp/armeria/server/Http1ServerDelayedCloseConnectionTest.java

```diff
@@ -0,0 +1,150 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.awaitility.Awaitility.await;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.net.BindException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.time.Duration;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.linecorp.armeria.common.Flags;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+class Http1ServerDelayedCloseConnectionTest {
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) {
+            sb.idleTimeoutMillis(0);
+            sb.http(0);
+            sb.https(0);
+            sb.tlsSelfSigned();
+            sb.service("/close", (ctx, req) -> {
+                return HttpResponse.builder()
+                                   .ok()
+                                   .content("OK\n")
+                                   .header(HttpHeaderNames.CONNECTION, "close")
+                                   .build();
+            });
+        }
+    };
+
+    @Test
+    void shouldDelayDisconnectByServerSideIfClientDoesNotHandleConnectionClose() throws IOException {
+        try (Socket socket = new Socket("127.0.0.1", server.httpPort())) {
+            socket.setSoTimeout(100000);
+            final int socketPort = socket.getLocalPort();
+            final PrintWriter writer = new PrintWriter(socket.getOutputStream());
+            writer.print("GET /close" + " HTTP/1.1\r\n");
+            writer.print("\r\n");
+            writer.flush();
+
+            final BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+            assertThat(in.readLine()).isEqualTo("HTTP/1.1 200 OK");
+
+            String line;
+            boolean hasConnectionClose = false;
+            while ((line = in.readLine()) != null) {
+                if ("connection: close".equalsIgnoreCase(line)) {
+                    hasConnectionClose = true;
+                }
+                if (line.isEmpty() || line.contains(":")) {
+                    continue;
+                }
+                if (line.startsWith("OK")) {
+                    break;
+                }
+            }
```

Comment:
nit, can we just explicitly check each line?

```suggestion
            final BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            assertThat(in.readLine()).isEqualTo("HTTP/1.1 200 OK");
            in.readLine(); // content-type
            in.readLine(); // content-length
            in.readLine(); // server
            in.readLine(); // date
            assertThat(in.readLine()).isEqualToIgnoringCase("connection: close");
            assertThat(in.readLine()).isEmpty();
            assertThat(in.readLine()).isEqualToIgnoringCase("OK");
```

Reply:
I refactored it too 🙇